### PR TITLE
refactor config reload in daemon

### PR DESCRIPTION
**Pull request by allencloud**
Signed-off-by: allencloud <allen.sun@daocloud.io>

With the code base growing up in aspect of reload, it seems that code becomes some kind of unreadable. This PR aimed at refactoring the reload. 
 
**- What I did**
1. split func reload into a single reload.go
2. split reload test from daemon_test.go to reload_test.go
3. refactor func Reload into several independent functions.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/daemon_solaris.go [hunk 1]
```diff
@@ -301,9 +301,8 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 	return warnings, nil
 }
 
-// platformReload updates configuration with platform specific options
-func (daemon *Daemon) platformReload(config *Config) map[string]string {
-	return map[string]string{}
+// reloadPlatform updates configuration with platform specific options
+func (daemon *Daemon) reloadPlatform(config *Config, attributes map[string]string) {
```
Comment (AkihiroSuda) [inline daemon/daemon_solaris.go]: What's the intention of this change?

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -577,10 +577,8 @@ func (daemon *Daemon) platformReload(config *Config) map[string]string {
 		runtimeList.WriteString(fmt.Sprintf("%s:%s", name, rt))
 	}
 
-	return map[string]string{
-		"runtimes":        runtimeList.String(),
-		"default-runtime": daemon.configStore.DefaultRuntime,
-	}
+	attributes["runtimes"] = runtimeList.String()
+	attributes["default-runtime"] = daemon.configStore.DefaultRuntime
```
Comment (AkihiroSuda) [inline daemon/daemon_unix.go]: What's the intention of this change?

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -556,8 +556,8 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 	return warnings, nil
 }
 
-// platformReload updates configuration with platform specific options
-func (daemon *Daemon) platformReload(config *Config) map[string]string {
+// reloadPlatform updates configuration with platform specific options
+func (daemon *Daemon) reloadPlatform(config *Config, attributes map[string]string) {
```
Comment (AkihiroSuda) [inline daemon/daemon_unix.go]: Please add a comment that callee updates the `attributes` ?

Generally I don't prefer args to be written.
An alternative design I can come up with is to keep returning `map`, and compose `map`s to a single `map` with some utility function.

i.e.
```go
mb := mapBuilder()
mb.Append(daemon.reloadPlatform(config))
mb.Append(daemon.reloadBlahBlahBlah(config))
attributes := mb.Map()
```

But maybe too complicated.

File: daemon/daemon_solaris.go (hunk 1 above)
Comment (allencloud) [inline daemon/daemon_solaris.go]: Here is my thought.
1. everything reload func has a name with a prefix `reload` as `reloadXxxx`, so I changed the function name
2. In my implementation, it needs to modify `attributes`, then it takes one more parameter, but with no return.

File: daemon/daemon_unix.go (hunk 3 above)
Comment (allencloud) [inline daemon/daemon_unix.go]: What you mentioned is what I am concerned with as well. I hope that we could add a specific comments as your suggested. Updated this PR. PTAL

**Review by LK4D4 — APPROVED** [review]
LGTM

**Comment by LK4D4** [comment]
@allencloud meh, need rebase :/

**Comment by allencloud** [comment]
Oh, Thanks @LK4D4 
Now rebased, PTAL 😄

**Comment by allencloud** [comment]
ping @AkihiroSuda @vdemeester @thaJeztah @aaronlehmann 
Could you help to review this one? ^^
Thanks a lot and feel free to ask me if I can provide more details.

**Comment by vdemeester** [comment]
hum @allencloud we're gonna be in conflicts you and me :joy:, see #29673. How my PR would affect yours ?

**Comment by allencloud** [comment]
Oh, thanks for your foresighted thought. I will come to check it so that we can make less conflict. @vdemeester 😄

**Comment by allencloud** [comment]
I think there are some affects there, @vdemeester :
1. `platformReload` function has conflict;
2. one must rebase if the other is merged. While I think rebase and change will not be so tough, but need to be careful.

**Comment by allencloud** [comment]
I have no idea why the CI is still running after about several hours. While when I enter the details of each CI test, the test shows SUCCESS. like janky:
```
Finished: SUCCESS
```
Is there some issue about CI? @thaJeztah @vdemeester

**Comment by cpuguy83** [comment]
Yeah, we're having CI issues.

**Comment by allencloud** [comment]
Oh, lucky to see all green. Could you please help to review? 😄  @vdemeester @cpuguy83 @LK4D4 @aaronlehmann

File: daemon/reload.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	if daemon.clusterProvider != nil {
+		if err := conf.IsSwarmCompatible(); err != nil {
+			return err
+		}
+	}
+
+	// check discovery modifications
+	if !config.ModifiedDiscoverySettings(daemon.configStore, newAdvertise, newClusterStore, conf.ClusterOpts) {
```
Comment (vdemeester) [inline daemon/reload.go]: There is a conflict on that line, it should be `if !config.ModifiedDiscoverySettings(daemon.configStore, newClusterStore, newAdvertise, conf.ClusterOpts) {`

**Review by vdemeester — APPROVED** [review]
LGTM 🦁
@allencloud Needs a rebase

**Comment by vieux** [comment]
LGTM

**Pull request merged**